{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Common URL API: https://ncbi.github.io/blast-cloud/dev/api.html  \n",
    "Biopython BLAST code: https://github.com/biopython/biopython/blob/171697883aca6894f8367f8f20f1463ce7784d0c/Bio/Blast/NCBIWWW.py  \n",
    "Useful comments: https://github.com/biopython/biopython/blob/a338d04074585afe7e1f8f418c9107c05dd5e38d/Doc/Tutorial/chapter_blast.tex  \n",
    "Example use: http://fenyolab.org/presentations/Introduction_Biostatistics_Bioinformatics_2014/tutorials/week3/BLAST%20with%20BioPython.pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copyright 1999 by Jeffrey Chang.  All rights reserved.\n",
    "#\n",
    "# This file is part of the Biopython distribution and governed by your\n",
    "# choice of the \"Biopython License Agreement\" or the \"BSD 3-Clause License\".\n",
    "# Please see the LICENSE file that should have been included as part of this\n",
    "# package.\n",
    "#\n",
    "# Patched by Brad Chapman.\n",
    "# Chris Wroe added modifications for work in myGrid\n",
    "\n",
    "\"\"\"\n",
    "Code to invoke the NCBI BLAST server over the internet.\n",
    "This module provides code to work with the WWW version of BLAST\n",
    "provided by the NCBI. https://blast.ncbi.nlm.nih.gov/\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext blackcellmagic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "import sys\n",
    "\n",
    "from urllib.request import urlopen\n",
    "from urllib.parse import urlencode\n",
    "from urllib.request import Request\n",
    "import requests\n",
    "\n",
    "import re\n",
    "from bs4 import BeautifulSoup\n",
    "from bs4 import Comment\n",
    "\n",
    "def blast(\n",
    "    program,\n",
    "    database,\n",
    "    sequence,\n",
    "    ncbi_gi=False,\n",
    "    descriptions=500,\n",
    "    alignments=500,\n",
    "    hitlist_size=50,\n",
    "    expect=10.0,\n",
    "    low_comp_filt=False,\n",
    "    megablast=True,\n",
    "    format_type=\"XML\",\n",
    "):\n",
    "    \"\"\"\n",
    "    BLAST search using NCBI's QBLAST server.\n",
    "    Supports all parameters of the old qblast API for Put (https://ncbi.github.io/blast-cloud/dev/api.html).\n",
    "    Please note that NCBI uses the new Common URL API for BLAST searches\n",
    "    on the internet (http://ncbi.github.io/blast-cloud/dev/api.html). Thus,\n",
    "    some of the parameters used by this function are not (or are no longer)\n",
    "    officially supported by NCBI. Although they are still functioning, this\n",
    "    may change in the future.\n",
    "\n",
    "    Some useful parameters:\n",
    "     - program        blastn, blastp, blastx, tblastn, or tblastx\n",
    "     - database       nt, nr, refseq_rna, refseq_protein, swissprot, pdbaa, or pdbnt\n",
    "                      (More info: https://ncbi.github.io/blast-cloud/blastdb/available-blastdbs.html)\n",
    "     - sequence       The sequence to search.\n",
    "     - ncbi_gi        True/False whether to return NCBI GI identifiers. Default False.\n",
    "     - descriptions   int or None. Limit number of descriptions to show. Default 500.\n",
    "     - alignments     int or None. Limit number of alignments to show. Default 500.\n",
    "     - hitlist_size   int or None. Limit number of hits to return. Default 50.\n",
    "     - expect         int or None. An expect value cutoff. Default 10.0.\n",
    "     - low_comp_filt  True/False whether to low complexity filter. Default False.\n",
    "     - megablast      True/False whether to use the MEGABLAST algorithm (blastn only). Default True.\n",
    "     - format_type    Return format. \"HTML\", \"Text\", \"ASN.1\", or \"XML\".  Default \"XML\".\n",
    "\n",
    "    This function does not check the validity of the parameters\n",
    "    and passes the values to the server as is.  More help is available at:\n",
    "    https://ncbi.github.io/blast-cloud/dev/api.html\n",
    "    \"\"\"\n",
    "    # Server rules:\n",
    "    # 1. Do not contact the server more often than once every 10 seconds.\n",
    "    # 2. Do not poll for any single RID more often than once a minute.\n",
    "    # 3. Use the URL parameter email and tool, so that the NCBI\n",
    "    #    can contact you if there is a problem.\n",
    "    # 4. Run scripts weekends or between 9 pm and 5 am Eastern time\n",
    "    #    on weekdays if more than 50 searches will be submitted.\n",
    "    # Reference: https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=DeveloperInfo\n",
    "    \n",
    "    # Define server URL and content type\n",
    "    url_base = \"https://blast.ncbi.nlm.nih.gov/Blast.cgi\"\n",
    "    content_type = \"application/x-www-form-urlencoded\"\n",
    "    client = \"ggetClient\"\n",
    "    \n",
    "    ## Clean up parameters\n",
    "    # If the path to afasta file was provided as sequence,\n",
    "    # read the file and extract the sequence\n",
    "    if \".fa\" in sequence:\n",
    "        from Bio import SeqIO\n",
    "        sequence = SeqIO.read(sequence, format=\"fasta\").seq\n",
    "\n",
    "    # Convert program to lower case\n",
    "    program = program.lower()\n",
    "    # Check if programs was defined as expected\n",
    "    programs = [\"blastn\", \"blastp\", \"blastx\", \"tblastn\", \"tblastx\"]\n",
    "    if program not in programs:\n",
    "        raise ValueError(\n",
    "            \"Program specified is %s. Expected one of %s\"\n",
    "            % (program, \", \".join(programs))\n",
    "        )\n",
    "    \n",
    "    # Translate filter and ncbi_gi parameters\n",
    "    if low_comp_filt == False:\n",
    "        low_comp_filt = None\n",
    "    else:\n",
    "        low_comp_filt = \"T\"\n",
    "\n",
    "    if ncbi_gi == False:\n",
    "        ncbi_gi = None\n",
    "    else:\n",
    "        ncbi_gi = \"T\"\n",
    "\n",
    "    if megablast == False:\n",
    "        megablast = None\n",
    "    else:\n",
    "        megablast = \"on\"\n",
    "        \n",
    "    parameters = [\n",
    "        (\"PROGRAM\", program),\n",
    "        (\"DATABASE\", database),\n",
    "        (\"QUERY\", sequence),\n",
    "        (\"NCBI_GI\", ncbi_gi),\n",
    "        (\"DESCRIPTIONS\", descriptions),\n",
    "        (\"ALIGNMENTS\", alignments),\n",
    "        (\"HITLIST_SIZE\", hitlist_size),\n",
    "        (\"EXPECT\", expect),\n",
    "        (\"FILTER\", low_comp_filt),\n",
    "        (\"MEGABLAST\", megablast),\n",
    "        (\"CMD\", \"Put\"),\n",
    "    ]\n",
    "    \n",
    "    ## Define query\n",
    "    query = [x for x in parameters if x[1] is not None]\n",
    "    message = urlencode(query).encode()\n",
    "\n",
    "    ## Submit search\n",
    "    request = Request(url_base, message, {\"User-Agent\": client})\n",
    "    handle = urlopen(request)\n",
    "\n",
    "    # Fetch Request ID (RID) and estimated time to completion (RTOE)\n",
    "    RID, RTOE = _parse_qblast_ref_page(handle)\n",
    "    sys.stderr.write(f\"Request ID: {RID}.\\n\")\n",
    "    \n",
    "    # Wait for search to complete\n",
    "    # (At least 10 seconds to comply with server rule 1.)\n",
    "    sys.stderr.write(f\"Estimated time to completion: {RTOE} seconds.\\n\")\n",
    "    if RTOE < 10:\n",
    "        time.sleep(10)\n",
    "    else:\n",
    "        time.sleep(int(RTOE))\n",
    "        \n",
    "    ## Poll NCBI until the results are ready\n",
    "    status_query = \"https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&FORMAT_OBJECT=SearchInfo&RID=\"\n",
    "    searching = True\n",
    "    i = 0\n",
    "    while searching == True:\n",
    "        if i != 0:\n",
    "            # Sleep for 65 seconds if first fetch was not succesful\n",
    "            # to comply with server rules\n",
    "            time.sleep(65)\n",
    "        \n",
    "        # Query for search status\n",
    "        r_status = requests.get(\n",
    "            status_query + RID, \n",
    "            headers={\n",
    "                \"Content-Type\" : content_type,\n",
    "                \"User-agent\": client\n",
    "            }\n",
    "        )\n",
    "        # Raise errors if status code is not 200\n",
    "        if r_status.status_code != 200:\n",
    "            raise RuntimeError(f\"HTTPS request response status code {r_status.status_code}. Please try again.\\n\")\n",
    "\n",
    "        # Get search status\n",
    "        soup_status = BeautifulSoup(r_status.content, \"html.parser\")\n",
    "        comments = soup_status.find_all(string=lambda text: isinstance(text, Comment))\n",
    "        status = [i for i in comments if i.startswith('\\n                QBlastInfoBegin')][0].split(\"\\n\")[2].split()[0]\n",
    "\n",
    "        if status == \"Status=WAITING\":\n",
    "            sys.stderr.write(\"Searching...\\n\")\n",
    "            i = 1\n",
    "            continue\n",
    "\n",
    "        if status == \"Status=FAILED\":\n",
    "            sys.stderr.write(f\"Search {RID} failed; please report to blast-help@ncbi.nlm.nih.gov.\\n\")\n",
    "            searching = False\n",
    "\n",
    "        if status == \"Status=UNKNOWN\":\n",
    "            sys.stderr.write(f\"Search {RID} expired.\\n\")\n",
    "            searching = False  \n",
    "\n",
    "        if status == \"Status=READY\":\n",
    "            sys.stderr.write(\"Search complete, retrieving results...\\n\")\n",
    "            searching = False\n",
    "\n",
    "        else:\n",
    "            sys.stderr.write(f\"\"\"\n",
    "            Something unexpected happened. \\n\n",
    "            Search {RID} possibly failed; please report to blast-help@ncbi.nlm.nih.gov\\n\n",
    "            or post an issue on Github: https://github.com/lauraluebbert/gget\\n\n",
    "            \"\"\")\n",
    "            searching = False\n",
    "        \n",
    "    ## Retrieve search results\n",
    "    results_query = \"https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&FORMAT_TYPE=Text&RID=\";\n",
    "\n",
    "    r_results = requests.get(\n",
    "        results_query + RID, \n",
    "        headers={\n",
    "            \"Content-Type\" : content_type,\n",
    "            \"User-agent\": client\n",
    "        }\n",
    "    )\n",
    "    # Raise errors if status code is not 200\n",
    "    if r_results.status_code != 200:\n",
    "        raise RuntimeError(f\"HTTPS request response status code {r_results.status_code}. Please try again.\\n\")\n",
    "        \n",
    "    soup = BeautifulSoup(r_results.content, \"html.parser\")\n",
    "    \n",
    "    return soup.find('pre')\n",
    "\n",
    "def _parse_qblast_ref_page(handle):\n",
    "    \"\"\"\n",
    "    Extract a tuple of RID, RTOE from the NCBI 'please wait' page.\n",
    "    RTOE = 'Estimated time fo completion.' \n",
    "    RID = 'Request ID'.\n",
    "    \"\"\"\n",
    "    s = handle.read().decode()\n",
    "    i = s.find(\"RID =\")\n",
    "    if i == -1:\n",
    "        rid = None\n",
    "    else:\n",
    "        j = s.find(\"\\n\", i)\n",
    "        rid = s[i + len(\"RID =\") : j].strip()\n",
    "\n",
    "    i = s.find(\"RTOE =\")\n",
    "    if i == -1:\n",
    "        rtoe = None\n",
    "    else:\n",
    "        j = s.find(\"\\n\", i)\n",
    "        rtoe = s[i + len(\"RTOE =\") : j].strip()\n",
    "\n",
    "    if not rid and not rtoe:\n",
    "        # Can we reliably extract the error message from the HTML page?\n",
    "        # e.g.  \"Message ID#24 Error: Failed to read the Blast query:\n",
    "        #       Nucleotide FASTA provided for protein sequence\"\n",
    "        # or    \"Message ID#32 Error: Query contains no data: Query\n",
    "        #       contains no sequence data\"\n",
    "        #\n",
    "        # This used to occur inside a <div class=\"error msInf\"> entry:\n",
    "        i = s.find('<div class=\"error msInf\">')\n",
    "        if i != -1:\n",
    "            msg = s[i + len('<div class=\"error msInf\">') :].strip()\n",
    "            msg = msg.split(\"</div>\", 1)[0].split(\"\\n\", 1)[0].strip()\n",
    "            if msg:\n",
    "                raise ValueError(\"Error message from NCBI: %s\" % msg)\n",
    "        # In spring 2010 the markup was like this:\n",
    "        i = s.find('<p class=\"error\">')\n",
    "        if i != -1:\n",
    "            msg = s[i + len('<p class=\"error\">') :].strip()\n",
    "            msg = msg.split(\"</p>\", 1)[0].split(\"\\n\", 1)[0].strip()\n",
    "            if msg:\n",
    "                raise ValueError(\"Error message from NCBI: %s\" % msg)\n",
    "        # Generic search based on the way the error messages start:\n",
    "        i = s.find(\"Message ID#\")\n",
    "        if i != -1:\n",
    "            # Break the message at the first HTML tag\n",
    "            msg = s[i:].split(\"<\", 1)[0].split(\"\\n\", 1)[0].strip()\n",
    "            raise ValueError(\"Error message from NCBI: %s\" % msg)\n",
    "        # If we cannot recognise the error layout:\n",
    "        raise ValueError(\n",
    "            \"No RID and no RTOE found in the 'please wait' page, \"\n",
    "            \"there was probably an error in your request but we \"\n",
    "            \"could not extract a helpful error message.\"\n",
    "        )\n",
    "    elif not rid:\n",
    "        # Can this happen?\n",
    "        raise ValueError(\n",
    "            \"No RID found in the 'please wait' page. (Although RTOE = %r)\" % rtoe\n",
    "        )\n",
    "    elif not rtoe:\n",
    "        # Can this happen?\n",
    "        raise ValueError(\n",
    "            \"No RTOE found in the 'please wait' page. (Although RID = %r)\" % rid\n",
    "        )\n",
    "\n",
    "    try:\n",
    "        return rid, int(rtoe)\n",
    "    except ValueError:\n",
    "        raise ValueError(\n",
    "            \"A non-integer RTOE found in the 'please wait' page, %r\" % rtoe\n",
    "        ) from None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "program = \"blastn\"\n",
    "db = \"nt\"\n",
    "\n",
    "seq = \"seq_results.fa\"\n",
    "# seq = \"TTTGTAGTTACATAGCAAAATGCGCGTTTATTTCGGCTCAGTAAATTAAGAACATTTTCGTTACACGTTGCCACGCCCCCCCATCCAACGGCAAACCACCCCCCGCCTCCGCCTTCAAGACAGAGAGAGAGATAGAATAAAACAGAGAGAGAGCAACAACAACAATATATAACCAAAACGAAAGATTTCCGCAAGAAACCACAAAAAAAAACAACAATTGCTTTGGTTCTGATGATTCTCGTAACAGCAAAAACAACAAGATTTATAATCTAAATCAAAAGAAACTGTGCTCTCCGCTGTGCGTGGCTGTGAGTGCGTTTGTGCGGTACGGTGCGTGCATTATAACAATTATTGCACAAAAGGCATGACAAACGTGGGGCGGGGGTGTCTGTAAATGTGAGTGCGATAGAGCGCGACATATTCAAACAAGCTAGGGACGGAGTAGGTGAGCGAGGGAGAGAGAGAACGCGACGGTGGCAACAACAAAGGGCTGTGATTTTACTTTGCCCCCTCCACCCTGTACCCCGTCCCATCCATACACTTGCGTATTTACCAAATAAACGGAACATACATAAGCCAAAATAGGCAGAACAACAACAACATTAAGAGTCAACTAAACAGCAAATCGAATAGTTTCGTGGGAGAAAGGACAGCGTAGCAGAGTCATTGGAAAACTGGCCGCGTAAGGGAGCGCAATGGATGTCCTGGAAATGCTGCGCGCCAGCGCCAGCGGCAGCTACAACACAATATTCTCGGACGCGTGGTGCCAATATGTCTCCAAGCAGATCACAGCAACGGTAAGTGCACTTCACCCTTTCTTTGTCTTTTTATTTGCTTTGACTTTGGCTTGCTGTTTGTTTGCAATATGAGTGAGTCAGTTGGAAGACAATATGTGCGAGAGCGAGAGGA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "blast(program, db, seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
