{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "from urllib.request import urlopen, Request\n",
    "from urllib.parse import urlencode\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import urllib\n",
    "from io import StringIO\n",
    "from IPython.display import display, HTML\n",
    "\n",
    "def n_colors(nucleotide):\n",
    "    \"\"\"\n",
    "    Returns a string format to print the nucleotide\n",
    "    with its appropriate background color according to the Clustal Colour Scheme.\n",
    "    \"\"\"\n",
    "\n",
    "    # Raw python background colors\n",
    "    # References:\n",
    "    # https://stackabuse.com/how-to-print-colored-text-in-python/\n",
    "    # https://www.ditig.com/256-colors-cheat-sheet\n",
    "    raw_colors = {\n",
    "        \"black\": 0,\n",
    "        \"red\": 9,\n",
    "        \"green\": 10,\n",
    "        \"yellow\": 11,\n",
    "        \"blue\": 12,\n",
    "        \"white\": 15,\n",
    "    }\n",
    "\n",
    "    # Define nulceotide colors dict\n",
    "    n_colors_dict = {\n",
    "        \"blue\": [\"C\"],\n",
    "        \"red\": [\"A\"],\n",
    "        \"green\": [\"T\"],\n",
    "        \"yellow\": [\"G\"],\n",
    "    }\n",
    "\n",
    "    # Define background color based on which nucleotide was passed\n",
    "    bkg_color = None\n",
    "    letter_color = None\n",
    "    for color in n_colors_dict:\n",
    "        if nucleotide in n_colors_dict[color]:\n",
    "            bkg_color = raw_colors[color]\n",
    "            letter_color = color\n",
    "\n",
    "    # If the nucleotide does not fall into the defined color categories,\n",
    "    # make it white (e.g. \"-\")\n",
    "    if bkg_color == None:\n",
    "        bkg_color = raw_colors[\"white\"]\n",
    "\n",
    "    if letter_color is not None and letter_color in [\"blue\", \"red\"]:\n",
    "        # Define textcolor as white for darker colors\n",
    "        textcolor = raw_colors[\"white\"]\n",
    "    else:\n",
    "        # Define textcolor as black\n",
    "        textcolor = raw_colors[\"black\"]\n",
    "\n",
    "    return f\"\\033[38;5;{textcolor}m\\033[48;5;{bkg_color}m{nucleotide}\\033[0;0m\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def aa_colors(amino_acid):\n",
    "    \"\"\"\n",
    "    Returns a string format to print the amino acid\n",
    "    with its appropriate background color according to the Clustal Colour Scheme:\n",
    "    http://www.jalview.org/help/html/colourSchemes/clustal.html\n",
    "    \"\"\"\n",
    "\n",
    "    # Raw python background colors\n",
    "    # References:\n",
    "    # https://stackabuse.com/how-to-print-colored-text-in-python/\n",
    "    # https://www.ditig.com/256-colors-cheat-sheet\n",
    "    raw_colors = {\n",
    "        \"black\": 0,\n",
    "        \"red\": 9,\n",
    "        \"green\": 10,\n",
    "        \"yellow\": 11,\n",
    "        \"blue\": 12,\n",
    "        \"cyan\": 14,\n",
    "        \"magenta\": 5,\n",
    "        \"pink\": 13,\n",
    "        \"orange\": 1,  # This is maroon because the system colors don't have orange\n",
    "        \"white\": 15,\n",
    "    }\n",
    "\n",
    "    # Define amino acid colors dict according to http://www.jalview.org/help/html/colourSchemes/clustal.html\n",
    "    aa_colors_dict = {\n",
    "        \"blue\": [\"A\", \"I\", \"L\", \"M\", \"F\", \"W\", \"V\", \"C\"],  # hydrophobic AAs\n",
    "        \"red\": [\"K\", \"R\"],  # positive charge\n",
    "        \"magenta\": [\"E\", \"D\"],  # negative charge\n",
    "        \"green\": [\"N\", \"Q\", \"S\", \"T\"],  # polar\n",
    "        \"pink\": [\"C\"],  # cysteines\n",
    "        \"orange\": [\"G\"],  # glycines\n",
    "        \"yellow\": [\"P\"],  # prolines\n",
    "        \"cyan\": [\"H\", \"Y\"],  # aromatic\n",
    "    }\n",
    "\n",
    "    # Define background color based on which amino acid was passed\n",
    "    bkg_color = None\n",
    "    letter_color = None\n",
    "    for color in aa_colors_dict:\n",
    "        if amino_acid in aa_colors_dict[color]:\n",
    "            bkg_color = raw_colors[color]\n",
    "            letter_color = color\n",
    "\n",
    "    # If the amino acid does not fall into the defined color categories,\n",
    "    # make it white (e.g. \"-\")\n",
    "    if bkg_color == None:\n",
    "        bkg_color = raw_colors[\"white\"]\n",
    "\n",
    "    if letter_color is not None and letter_color in [\n",
    "        \"blue\",\n",
    "        \"red\",\n",
    "        \"magenta\",\n",
    "        \"orange\",\n",
    "    ]:\n",
    "        # Define textcolor as white for darker colors\n",
    "        textcolor = raw_colors[\"white\"]\n",
    "    else:\n",
    "        # Define textcolor as black\n",
    "        textcolor = raw_colors[\"black\"]\n",
    "\n",
    "    return f\"\\033[38;5;{textcolor}m\\033[48;5;{bkg_color}m{amino_acid}\\033[0;0m\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\x1b[38;5;15m\\x1b[48;5;12mL\\x1b[0;0m'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_colors(\"L\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_uniprot_seqs(server, ensembl_ids):\n",
    "    \"\"\"\n",
    "    Retrieve UniProt sequences based on Ensemsbl identifiers.\n",
    "\n",
    "    Args:\n",
    "    - server\n",
    "    Link to UniProt REST API server.\n",
    "    - ensembl_ids:\n",
    "    One or more transcript Ensembl IDs (string or list of strings).\n",
    "\n",
    "    Returns data frame with UniProt ID, gene name, organism, sequence, sequence length, and query ID.\n",
    "    \"\"\"\n",
    "\n",
    "    # If a single UniProt ID is passed as string, convert to list\n",
    "    if type(ensembl_ids) == str:\n",
    "        ensembl_ids = [ensembl_ids]\n",
    "\n",
    "    # Define query arguments\n",
    "    # Columns documentation: https://www.uniprot.org/help/uniprotkb%5Fcolumn%5Fnames\n",
    "    # from/to IDs documentation: https://www.uniprot.org/help/api_idmapping\n",
    "    query_args = {\n",
    "        \"from\": \"ENSEMBL_TRS_ID\",\n",
    "        \"to\": \"ACC\",\n",
    "        \"format\": \"tab\",\n",
    "        \"query\": \" \".join(ensembl_ids),\n",
    "        \"columns\": \"id,genes,organism,sequence,length\",\n",
    "    }\n",
    "    # Reformat query arguments\n",
    "    query_args = urllib.parse.urlencode(query_args)\n",
    "    query_args = query_args.encode(\"ascii\")\n",
    "\n",
    "    # Submit query to UniProt server\n",
    "    request = urllib.request.Request(server, query_args)\n",
    "\n",
    "    # Read and clean up results\n",
    "    with urllib.request.urlopen(request) as response:\n",
    "        res = response.read()\n",
    "    \n",
    "    # Check if URL retruned error code\n",
    "    if response.getcode() != 200:\n",
    "        raise RuntimeError(\n",
    "            f\"The UniProt server returned error status code {response.getcode()}. Please try again.\"\n",
    "        )\n",
    "    # Initiate data frame so empty df will be returned if no matches are found\n",
    "    df = pd.DataFrame()\n",
    "\n",
    "    try:\n",
    "        # This will throw an EmptyDataError if no results were found\n",
    "        df = pd.read_csv(StringIO(res.decode(\"utf-8\")), sep=\"\\t\")\n",
    "\n",
    "        if len(df.columns) == 6:\n",
    "            # Rename columns\n",
    "            df.columns = [\n",
    "                \"uniprot_id\",\n",
    "                \"gene_name\",\n",
    "                \"organism\",\n",
    "                \"sequence\",\n",
    "                \"sequence_length\",\n",
    "                \"query\",\n",
    "            ]\n",
    "\n",
    "        # Sometimes a seventh \"isomap\" column is returned.\n",
    "        if len(df.columns) == 7:\n",
    "            # Drop isoform column (last column)\n",
    "            df = df.iloc[:, :-1]\n",
    "            # Rename columns\n",
    "            df.columns = [\n",
    "                \"uniprot_id\",\n",
    "                \"gene_name\",\n",
    "                \"organism\",\n",
    "                \"sequence\",\n",
    "                \"sequence_length\",\n",
    "                \"query\",\n",
    "            ]\n",
    "\n",
    "        # Split rows if two different UniProt IDs for a single query ID are returned\n",
    "        df = df.assign(Query=df[\"query\"].str.split(\",\")).explode(\"query\")\n",
    "\n",
    "    # If no results were found, do nothing (returns the empty data frame)\n",
    "    except pd.errors.EmptyDataError:\n",
    "        None\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_uniprot_seqs(\"https://www.uniprot.org/uploadlists/\", \"banana\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = get_uniprot_seqs(\"https://www.uniprot.org/uploadlists/\", [\"ENST00000392653\", \"ENST00000392657\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['P35326',\n",
       "  'SPRR2A',\n",
       "  'Homo sapiens (Human)',\n",
       "  'MSYQQQQCKQPCQPPPVCPTPKCPEPCPPPKCPEPCPPPKCPQPCPPQQCQQKYPPVTPSPPCQSKYPPKSK',\n",
       "  72,\n",
       "  'ENST00000392653',\n",
       "  ['ENST00000392653']],\n",
       " ['A7KAX9',\n",
       "  'ARHGAP32 GRIT KIAA0712 RICS',\n",
       "  'Homo sapiens (Human)',\n",
       "  'METESESSTLGDDSVFWLESEVIIQVTDCEEEEREEKFRKMKSSVHSEEDDFVPELHRNVHPRERPDWEETLSAMARGADVPEIPGDLTLKTCGSTASMKVKHVKKLPFTKGHFPKMAECAHFHYENVEFGSIQLSLSEEQNEVMKNGCESKELVYLVQIACQGKSWIVKRSYEDFRVLDKHLHLCIYDRRFSQLSELPRSDTLKDSPESVTQMLMAYLSRLSAIAGNKINCGPALTWMEIDNKGNHLLVHEESSINTPAVGAAHVIKRYTARAPDELTLEVGDIVSVIDMPPKVLSTWWRGKHGFQVGLFPGHCVELINQKVPQSVTNSVPKPVSKKHGKLITFLRTFMKSRPTKQKLKQRGILKERVFGCDLGEHLLNSGFEVPQVLQSCTAFIERYGIVDGIYRLSGVASNIQRLRHEFDSEHVPDLTKEPYVQDIHSVGSLCKLYFRELPNPLLTYQLYEKFSDAVSAATDEERLIKIHDVIQQLPPPHYRTLEFLMRHLSLLADYCSITNMHAKNLAIVWAPNLLRSKQIESACFSGTAAFMEVRIQSVVVEFILNHVDVLFSGRISMAMQEGAASLSRPKSLLVSSPSTKLLTLEEAQARTQAQVNSPIVTENKYIEVGEGPAALQGKFHTIIEFPLERKRPQNKMKKSPVGSWRSFFNLGKSSSVSKRKLQRNESEPSEMKAMALKGGRAEGTLRSAKSEESLTSLHAVDGDSKLFRPRRPRSSSDALSASFNGEMLGNRCNSYDNLPHDNESEEEGGLLHIPALMSPHSAEDVDLSPPDIGVASLDFDPMSFQCSPPKAESECLESGASFLDSPGYSKDKPSANKKDAETGSSQCQTPGSTASSEPVSPLQEKLSPFFTLDLSPTEDKSSKPSSFTEKVVYAFSPKIGRKLSKSPSMSISEPISVTLPPRVSEVIGTVSNTTAQNASSSTWDKCVEERDATNRSPTQIVKMKTNETVAQEAYESEVQPLDQVAAEEVELPGKEDQSVSSSQSKAVASGQTQTGAVTHDPPQDSVPVSSVSLIPPPPPPKNVARMLALALAESAQQASTQSLKRPGTSQAGYTNYGDIAVATTEDNLSSSYSAVALDKAYFQTDRPAEQFHLQNNAPGNCDHPLPETTATGDPTHSNTTESGEQHHQVDLTGNQPHQAYLSGDPEKARITSVPLDSEKSDDHVSFPEDQSGKNSMPTVSFLDQDQSPPRFYSGDQPPSYLGASVDKLHHPLEFADKSPTPPNLPSDKIYPPSGSPEENTSTATMTYMTTTPATAQMSTKEASWDVAEQPTTADFAAATLQRTHRTNRPLPPPPSQRSAEQPPVVGQVQAATNIGLNNSHKVQGVVPVPERPPEPRAMDDPASAFISDSGAAAAQCPMATAVQPGLPEKVRDGARVPLLHLRAESVPAHPCGFPAPLPPTRMMESKMIAAIHSSSADATSSSNYHSFVTASSTSVDDALPLPLPVPQPKHASQKTVYSSFARPDVTTEPFGPDNCLHFNMTPNCQYRPQSVPPHHNKLEQHQVYGARSEPPASMGLRYNTYVAPGRNASGHHSKPCSRVEYVSSLSSSVRNTCYPEDIPPYPTIRRVQSLHAPPSSMIRSVPISRTEVPPDDEPAYCPRPLYQYKPYQSSQARSDYHVTQLQPYFENGRVHYRYSPYSSSSSSYYSPDGALCDVDAYGTVQLRPLHRLPNRDFAFYNPRLQGKSLYSYAGLAPRPRANVTGYFSPNDHNVVSMPPAADVKHTYTSWDLEDMEKYRMQSIRRESRARQKVKGPVMSQYDNMTPAVQDDLGGIYVIHLRSKSDPGKTGLLSVAEGKESRHAAKAISPEGEDRFYRRHPEAEMDRAHHHGGHGSTQPEKPSLPQKQSSLRSRKLPDMGCSLPEHRAHQEASHRQFCESKNGPPYPQGAGQLDYGSKGIPDTSEPVSYHNSGVKYAASGQESLRLNHKEVRLSKEMERPWVRQPSAPEKHSRDCYKEEEHLTQSIVPPPKPERSHSLKLHHTQNVERDPSVLYQYQPHGKRQSSVTVVSQYDNLEDYHSLPQHQRGVFGGGGMGTYVPPGFPHPQSRTYATALGQGAFLPAELSLQHPETQIHAE',\n",
       "  2087,\n",
       "  'ENST00000392657',\n",
       "  ['ENST00000392657']]]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/gd/8j_w3s656d1gkl7jfdnptqpc0000gq/T/ipykernel_17879/1171777585.py:1: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  df.values == [['P35326', 'SPRR2A', 'Homo sapiens (Human)',\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[ True,  True,  True,  True,  True,  True,  True],\n",
       "       [ True,  True,  True,  True,  True,  True,  True]])"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.values == [['P35326', 'SPRR2A', 'Homo sapiens (Human)',\n",
    "        'MSYQQQQCKQPCQPPPVCPTPKCPEPCPPPKCPEPCPPPKCPQPCPPQQCQQKYPPVTPSPPCQSKYPPKSK',\n",
    "        72, 'ENST00000392653', list(['ENST00000392653'])],\n",
    "       ['A7KAX9', 'ARHGAP32 GRIT KIAA0712 RICS', 'Homo sapiens (Human)',\n",
    "        'METESESSTLGDDSVFWLESEVIIQVTDCEEEEREEKFRKMKSSVHSEEDDFVPELHRNVHPRERPDWEETLSAMARGADVPEIPGDLTLKTCGSTASMKVKHVKKLPFTKGHFPKMAECAHFHYENVEFGSIQLSLSEEQNEVMKNGCESKELVYLVQIACQGKSWIVKRSYEDFRVLDKHLHLCIYDRRFSQLSELPRSDTLKDSPESVTQMLMAYLSRLSAIAGNKINCGPALTWMEIDNKGNHLLVHEESSINTPAVGAAHVIKRYTARAPDELTLEVGDIVSVIDMPPKVLSTWWRGKHGFQVGLFPGHCVELINQKVPQSVTNSVPKPVSKKHGKLITFLRTFMKSRPTKQKLKQRGILKERVFGCDLGEHLLNSGFEVPQVLQSCTAFIERYGIVDGIYRLSGVASNIQRLRHEFDSEHVPDLTKEPYVQDIHSVGSLCKLYFRELPNPLLTYQLYEKFSDAVSAATDEERLIKIHDVIQQLPPPHYRTLEFLMRHLSLLADYCSITNMHAKNLAIVWAPNLLRSKQIESACFSGTAAFMEVRIQSVVVEFILNHVDVLFSGRISMAMQEGAASLSRPKSLLVSSPSTKLLTLEEAQARTQAQVNSPIVTENKYIEVGEGPAALQGKFHTIIEFPLERKRPQNKMKKSPVGSWRSFFNLGKSSSVSKRKLQRNESEPSEMKAMALKGGRAEGTLRSAKSEESLTSLHAVDGDSKLFRPRRPRSSSDALSASFNGEMLGNRCNSYDNLPHDNESEEEGGLLHIPALMSPHSAEDVDLSPPDIGVASLDFDPMSFQCSPPKAESECLESGASFLDSPGYSKDKPSANKKDAETGSSQCQTPGSTASSEPVSPLQEKLSPFFTLDLSPTEDKSSKPSSFTEKVVYAFSPKIGRKLSKSPSMSISEPISVTLPPRVSEVIGTVSNTTAQNASSSTWDKCVEERDATNRSPTQIVKMKTNETVAQEAYESEVQPLDQVAAEEVELPGKEDQSVSSSQSKAVASGQTQTGAVTHDPPQDSVPVSSVSLIPPPPPPKNVARMLALALAESAQQASTQSLKRPGTSQAGYTNYGDIAVATTEDNLSSSYSAVALDKAYFQTDRPAEQFHLQNNAPGNCDHPLPETTATGDPTHSNTTESGEQHHQVDLTGNQPHQAYLSGDPEKARITSVPLDSEKSDDHVSFPEDQSGKNSMPTVSFLDQDQSPPRFYSGDQPPSYLGASVDKLHHPLEFADKSPTPPNLPSDKIYPPSGSPEENTSTATMTYMTTTPATAQMSTKEASWDVAEQPTTADFAAATLQRTHRTNRPLPPPPSQRSAEQPPVVGQVQAATNIGLNNSHKVQGVVPVPERPPEPRAMDDPASAFISDSGAAAAQCPMATAVQPGLPEKVRDGARVPLLHLRAESVPAHPCGFPAPLPPTRMMESKMIAAIHSSSADATSSSNYHSFVTASSTSVDDALPLPLPVPQPKHASQKTVYSSFARPDVTTEPFGPDNCLHFNMTPNCQYRPQSVPPHHNKLEQHQVYGARSEPPASMGLRYNTYVAPGRNASGHHSKPCSRVEYVSSLSSSVRNTCYPEDIPPYPTIRRVQSLHAPPSSMIRSVPISRTEVPPDDEPAYCPRPLYQYKPYQSSQARSDYHVTQLQPYFENGRVHYRYSPYSSSSSSYYSPDGALCDVDAYGTVQLRPLHRLPNRDFAFYNPRLQGKSLYSYAGLAPRPRANVTGYFSPNDHNVVSMPPAADVKHTYTSWDLEDMEKYRMQSIRRESRARQKVKGPVMSQYDNMTPAVQDDLGGIYVIHLRSKSDPGKTGLLSVAEGKESRHAAKAISPEGEDRFYRRHPEAEMDRAHHHGGHGSTQPEKPSLPQKQSSLRSRKLPDMGCSLPEHRAHQEASHRQFCESKNGPPYPQGAGQLDYGSKGIPDTSEPVSYHNSGVKYAASGQESLRLNHKEVRLSKEMERPWVRQPSAPEKHSRDCYKEEEHLTQSIVPPPKPERSHSLKLHHTQNVERDPSVLYQYQPHGKRQSSVTVVSQYDNLEDYHSLPQHQRGVFGGGGMGTYVPPGFPHPQSRTYATALGQGAFLPAELSLQHPETQIHAE',\n",
    "        2087, 'ENST00000392657', list(['ENST00000392657'])]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_uniprot_info(server, ensembl_ids, id_type):\n",
    "    \"\"\"\n",
    "    Retrieve UniProt synonyms and description based on Ensemsbl identifiers.\n",
    "\n",
    "    Args:\n",
    "    - server          Link to UniProt REST API server.\n",
    "    - ensembl_ids     One or more transcript Ensembl IDs (string or list of strings).\n",
    "    - id_type         \"Gene\" or \"Transcript\"\n",
    "\n",
    "    Returns data frame with UniProt ID, gene name, organism, sequence, sequence length, and query ID.\n",
    "    \"\"\"\n",
    "\n",
    "    # If a single Ensembl ID is passed as string, convert to list\n",
    "    if type(ensembl_ids) == str:\n",
    "        ensembl_ids = [ensembl_ids]\n",
    "\n",
    "    if id_type == \"Gene\":\n",
    "        ens_id_type = \"ENSEMBL_ID\"\n",
    "    elif id_type == \"Transcript\":\n",
    "        ens_id_type = \"ENSEMBL_TRS_ID\"\n",
    "    else:\n",
    "        logging.warning(\n",
    "            f\"Ensembl_ID '{ensembl_ids}' was not recognized as either gene nor transcript. Gene name synonyms and description will not be fetched from UniProt.\"\n",
    "        )\n",
    "        return\n",
    "\n",
    "    # Define query arguments\n",
    "    # Columns documentation: https://www.uniprot.org/help/uniprotkb%5Fcolumn%5Fnames\n",
    "    # from/to IDs documentation: https://www.uniprot.org/help/api_idmapping\n",
    "    query_args = {\n",
    "        \"from\": ens_id_type,\n",
    "        \"to\": \"ACC\",\n",
    "        \"format\": \"tab\",\n",
    "        \"query\": \" \".join(ensembl_ids),\n",
    "        \"columns\": \"id,genes(PREFERRED),genes,protein names,comment(FUNCTION)\",\n",
    "    }\n",
    "    # Reformat query arguments\n",
    "    query_args = urllib.parse.urlencode(query_args)\n",
    "    query_args = query_args.encode(\"ascii\")\n",
    "\n",
    "    # Submit query to UniProt server\n",
    "    request = urllib.request.Request(server, query_args)\n",
    "\n",
    "    # Read and clean up results\n",
    "    with urllib.request.urlopen(request) as response:\n",
    "        res = response.read()\n",
    "\n",
    "    # Check if URL retruned error code\n",
    "    if response.getcode() != 200:\n",
    "        raise RuntimeError(\n",
    "            f\"The UniProt server returned error status code {response.getcode()}. Please try again.\"\n",
    "        )\n",
    "\n",
    "    # Initiate data frame so empty df will be returned if no matches are found\n",
    "    df = pd.DataFrame()\n",
    "\n",
    "    try:\n",
    "        # This will throw an EmptyDataError if no results were found\n",
    "        df = pd.read_csv(StringIO(res.decode(\"utf-8\")), sep=\"\\t\")\n",
    "\n",
    "        if len(df.columns) == 6:\n",
    "            # Rename columns\n",
    "            df.columns = [\n",
    "                \"uniprot_id\",\n",
    "                \"primary_gene_name\",\n",
    "                \"synonyms\",\n",
    "                \"protein_names\",\n",
    "                \"uniprot_description\",\n",
    "                \"query\",\n",
    "            ]\n",
    "        # Sometimes a seventh \"isomap\" column is returned.\n",
    "        if len(df.columns) == 7:\n",
    "            # Drop isoform column (last column)\n",
    "            df = df.iloc[:, :-1]\n",
    "            # Rename columns\n",
    "            df.columns = [\n",
    "                \"uniprot_id\",\n",
    "                \"primary_gene_name\",\n",
    "                \"synonyms\",\n",
    "                \"protein_names\",\n",
    "                \"uniprot_description\",\n",
    "                \"query\",\n",
    "            ]\n",
    "        \n",
    "        try:\n",
    "            # Split gene names into list of strings\n",
    "            df[\"synonyms\"] = df[\"synonyms\"].str.split(\" \")\n",
    "        except:\n",
    "            None\n",
    "\n",
    "    # If no results were found, return None\n",
    "    except pd.errors.EmptyDataError:\n",
    "        return None\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([['A0A087XU86', nan, nan, 'Fish-egg lectin-like', nan,\n",
       "        'ENSPFOG00000009362'],\n",
       "       ['A0A8C5B536', 'LOC115543029', list(['LOC115543029']),\n",
       "        'Fish-egg lectin-like', nan, 'ENSGMOG00000033866'],\n",
       "       ['A0A8C5B611', 'LOC115543029', list(['LOC115543029']),\n",
       "        'Fish-egg lectin-like', nan, 'ENSGMOG00000033866'],\n",
       "       ['A0A8C5BQR4', 'LOC115543029', list(['LOC115543029']),\n",
       "        'Fish-egg lectin-like', nan, 'ENSGMOG00000033866'],\n",
       "       ['A0A8C5C537', 'LOC115543029', list(['LOC115543029']),\n",
       "        'Fish-egg lectin-like', nan, 'ENSGMOG00000033866'],\n",
       "       ['A0A8C5CE04', 'LOC115543029', list(['LOC115543029']),\n",
       "        'Fish-egg lectin-like', nan, 'ENSGMOG00000033866'],\n",
       "       ['A0A8C5FPH9', 'LOC115543029', list(['LOC115543029']),\n",
       "        'Fish-egg lectin-like', nan, 'ENSGMOG00000033866']], dtype=object)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_uniprot_info(\"https://www.uniprot.org/uploadlists/\", [\"ENSPFOG00000009362\", \"ENSGMOG00000033866\"], id_type=\"Gene\").values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([['A0A087XU86', nan, nan, 'Fish-egg lectin-like', nan,\n",
       "        'ENSPFOT00000009352'],\n",
       "       ['A0A8C5B536', 'LOC115543029', list(['LOC115543029']),\n",
       "        'Fish-egg lectin-like', nan, 'ENSGMOT00000059188']], dtype=object)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_uniprot_info(\"https://www.uniprot.org/uploadlists/\", [\"ENSPFOT00000009352\", \"ENSGMOT00000059188\"], id_type=\"Transcript\").values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_uniprot_info(\"https://www.uniprot.org/uploadlists/\", \"banana\", id_type=\"Transcript\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def wrap_cols_func(df, cols):\n",
    "    \"\"\"\n",
    "    Function to wrap columns cols of a\n",
    "    data frame df for easier reading.\n",
    "    \"\"\"\n",
    "    for col in cols:\n",
    "        df.loc[:, col] = df[col].str.wrap(30)\n",
    "\n",
    "    return display(HTML(df.to_html().replace(\"\\\\n\", \"<br>\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_column</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>This is a very very long text that needs to be...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         test_column\n",
       "0  This is a very very long text that needs to be..."
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame()\n",
    "df[\"test_column\"] = [\"This is a very very long text that needs to be wrapped because otherwise it will be very difficult to read and that will be quite inconvenient for the user of this beautiful program.\"]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_column</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>This is a very very long text<br>that needs to be wrapped<br>because otherwise it will be<br>very difficult to read and<br>that will be quite<br>inconvenient for the user of<br>this beautiful program.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "wrap_cols_func(df, [\"test_column\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rest_query(server, query, content_type):\n",
    "    \"\"\"\n",
    "    Function to query a\n",
    "\n",
    "    Args:\n",
    "    - server\n",
    "Server to query.\n",
    "    - Query\n",
    "    Query that is passed to server.\n",
    "    - content_type\n",
    "    Contect type requested from server.\n",
    "\n",
    "    Returns server output.\n",
    "    \"\"\"\n",
    "\n",
    "    r = requests.get(server + query, headers={\"Content-Type\": content_type})\n",
    "\n",
    "    if not r.ok:\n",
    "        raise RuntimeError(\n",
    "            f\"{server} returned error status code {r.status_code}. \"\n",
    "            \"Please double-check arguments and try again.\\n\"\n",
    "        )\n",
    "\n",
    "    if content_type == \"application/json\":\n",
    "        return r.json()\n",
    "    else:\n",
    "        return r.text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "http://rest.ensembl.org/ returned error status code 400. Please double-check arguments and try again.\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[1;32m/Users/lauraluebbert/OneDrive - California Institute of Technology/Data_analysis/gget/gget/test.ipynb Cell 18'\u001b[0m in \u001b[0;36m<cell line: 10>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/lauraluebbert/OneDrive%20-%20California%20Institute%20of%20Technology/Data_analysis/gget/gget/test.ipynb#ch0000016?line=5'>6</a>\u001b[0m ensembl_ID \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mbanana\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/lauraluebbert/OneDrive%20-%20California%20Institute%20of%20Technology/Data_analysis/gget/gget/test.ipynb#ch0000016?line=7'>8</a>\u001b[0m query \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mlookup/id/\u001b[39m\u001b[39m\"\u001b[39m \u001b[39m+\u001b[39m ensembl_ID \u001b[39m+\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m?\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/lauraluebbert/OneDrive%20-%20California%20Institute%20of%20Technology/Data_analysis/gget/gget/test.ipynb#ch0000016?line=9'>10</a>\u001b[0m df_temp \u001b[39m=\u001b[39m rest_query(server, query, content_type)\n",
      "\u001b[1;32m/Users/lauraluebbert/OneDrive - California Institute of Technology/Data_analysis/gget/gget/test.ipynb Cell 17'\u001b[0m in \u001b[0;36mrest_query\u001b[0;34m(server, query, content_type)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lauraluebbert/OneDrive%20-%20California%20Institute%20of%20Technology/Data_analysis/gget/gget/test.ipynb#ch0000015?line=15'>16</a>\u001b[0m r \u001b[39m=\u001b[39m requests\u001b[39m.\u001b[39mget(server \u001b[39m+\u001b[39m query, headers\u001b[39m=\u001b[39m{\u001b[39m\"\u001b[39m\u001b[39mContent-Type\u001b[39m\u001b[39m\"\u001b[39m: content_type})\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lauraluebbert/OneDrive%20-%20California%20Institute%20of%20Technology/Data_analysis/gget/gget/test.ipynb#ch0000015?line=17'>18</a>\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m r\u001b[39m.\u001b[39mok:\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/lauraluebbert/OneDrive%20-%20California%20Institute%20of%20Technology/Data_analysis/gget/gget/test.ipynb#ch0000015?line=18'>19</a>\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mRuntimeError\u001b[39;00m(\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lauraluebbert/OneDrive%20-%20California%20Institute%20of%20Technology/Data_analysis/gget/gget/test.ipynb#ch0000015?line=19'>20</a>\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mserver\u001b[39m}\u001b[39;00m\u001b[39m returned error status code \u001b[39m\u001b[39m{\u001b[39;00mr\u001b[39m.\u001b[39mstatus_code\u001b[39m}\u001b[39;00m\u001b[39m. \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lauraluebbert/OneDrive%20-%20California%20Institute%20of%20Technology/Data_analysis/gget/gget/test.ipynb#ch0000015?line=20'>21</a>\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mPlease double-check arguments and try again.\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lauraluebbert/OneDrive%20-%20California%20Institute%20of%20Technology/Data_analysis/gget/gget/test.ipynb#ch0000015?line=21'>22</a>\u001b[0m     )\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lauraluebbert/OneDrive%20-%20California%20Institute%20of%20Technology/Data_analysis/gget/gget/test.ipynb#ch0000015?line=23'>24</a>\u001b[0m \u001b[39mif\u001b[39;00m content_type \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mapplication/json\u001b[39m\u001b[39m\"\u001b[39m:\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/lauraluebbert/OneDrive%20-%20California%20Institute%20of%20Technology/Data_analysis/gget/gget/test.ipynb#ch0000015?line=24'>25</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m r\u001b[39m.\u001b[39mjson()\n",
      "\u001b[0;31mRuntimeError\u001b[0m: http://rest.ensembl.org/ returned error status code 400. Please double-check arguments and try again.\n"
     ]
    }
   ],
   "source": [
    "# Define Ensembl REST API server\n",
    "server = \"http://rest.ensembl.org/\"\n",
    "# Define type of returned content from REST\n",
    "content_type = \"application/json\"\n",
    "\n",
    "ensembl_ID = \"banana\"\n",
    "\n",
    "query = \"lookup/id/\" + ensembl_ID + \"?\"\n",
    "\n",
    "df_temp = rest_query(server, query, content_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': 'ENSPFOG00000009362',\n",
       " 'biotype': 'protein_coding',\n",
       " 'species': 'poecilia_formosa',\n",
       " 'assembly_name': 'PoeFor_5.1.2',\n",
       " 'description': 'fish-egg lectin-like [Source:NCBI gene;Acc:103152571]',\n",
       " 'object_type': 'Gene',\n",
       " 'start': 502511,\n",
       " 'end': 504269,\n",
       " 'Transcript': [{'db_type': 'core',\n",
       "   'source': 'ensembl',\n",
       "   'seq_region_name': 'KI519928.1',\n",
       "   'strand': 1,\n",
       "   'Translation': {'object_type': 'Translation',\n",
       "    'Parent': 'ENSPFOT00000009352',\n",
       "    'end': 503846,\n",
       "    'length': 264,\n",
       "    'start': 502638,\n",
       "    'version': 1,\n",
       "    'id': 'ENSPFOP00000009339',\n",
       "    'species': 'poecilia_formosa',\n",
       "    'db_type': 'core'},\n",
       "   'version': 1,\n",
       "   'logic_name': 'ensembl',\n",
       "   'Parent': 'ENSPFOG00000009362',\n",
       "   'assembly_name': 'PoeFor_5.1.2',\n",
       "   'species': 'poecilia_formosa',\n",
       "   'id': 'ENSPFOT00000009352',\n",
       "   'biotype': 'protein_coding',\n",
       "   'Exon': [{'assembly_name': 'PoeFor_5.1.2',\n",
       "     'species': 'poecilia_formosa',\n",
       "     'seq_region_name': 'KI519928.1',\n",
       "     'db_type': 'core',\n",
       "     'strand': 1,\n",
       "     'id': 'ENSPFOE00000114332',\n",
       "     'end': 502719,\n",
       "     'start': 502511,\n",
       "     'version': 1,\n",
       "     'object_type': 'Exon'},\n",
       "    {'end': 503105,\n",
       "     'start': 502869,\n",
       "     'version': 1,\n",
       "     'object_type': 'Exon',\n",
       "     'db_type': 'core',\n",
       "     'species': 'poecilia_formosa',\n",
       "     'assembly_name': 'PoeFor_5.1.2',\n",
       "     'seq_region_name': 'KI519928.1',\n",
       "     'strand': 1,\n",
       "     'id': 'ENSPFOE00000114345'},\n",
       "    {'strand': 1,\n",
       "     'id': 'ENSPFOE00000114457',\n",
       "     'species': 'poecilia_formosa',\n",
       "     'seq_region_name': 'KI519928.1',\n",
       "     'assembly_name': 'PoeFor_5.1.2',\n",
       "     'db_type': 'core',\n",
       "     'object_type': 'Exon',\n",
       "     'end': 503404,\n",
       "     'start': 503193,\n",
       "     'version': 1},\n",
       "    {'version': 1,\n",
       "     'start': 503490,\n",
       "     'end': 503614,\n",
       "     'object_type': 'Exon',\n",
       "     'species': 'poecilia_formosa',\n",
       "     'assembly_name': 'PoeFor_5.1.2',\n",
       "     'seq_region_name': 'KI519928.1',\n",
       "     'db_type': 'core',\n",
       "     'id': 'ENSPFOE00000114598',\n",
       "     'strand': 1},\n",
       "    {'start': 503708,\n",
       "     'version': 1,\n",
       "     'end': 504269,\n",
       "     'object_type': 'Exon',\n",
       "     'db_type': 'core',\n",
       "     'species': 'poecilia_formosa',\n",
       "     'assembly_name': 'PoeFor_5.1.2',\n",
       "     'seq_region_name': 'KI519928.1',\n",
       "     'id': 'ENSPFOE00000114741',\n",
       "     'strand': 1}],\n",
       "   'start': 502511,\n",
       "   'is_canonical': 1,\n",
       "   'end': 504269,\n",
       "   'object_type': 'Transcript'}],\n",
       " 'strand': 1,\n",
       " 'seq_region_name': 'KI519928.1',\n",
       " 'canonical_transcript': 'ENSPFOT00000009352.1',\n",
       " 'source': 'ensembl',\n",
       " 'db_type': 'core',\n",
       " 'version': 1,\n",
       " 'logic_name': 'ensembl'}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_latest_ens_rel():\n",
    "    \"\"\"\n",
    "    Returns the latest Ensembl release number.\n",
    "    \"\"\"\n",
    "    url = \"http://ftp.ensembl.org/pub/\"\n",
    "    html = requests.get(url)\n",
    "\n",
    "    # Raise error if status code not \"OK\" Response\n",
    "    if html.status_code != 200:\n",
    "        raise RuntimeError(\n",
    "            f\"The Ensembl FTP server returned error status code {html.status_code}. Please try again.\"\n",
    "        )\n",
    "\n",
    "    soup = BeautifulSoup(html.text, \"html.parser\")\n",
    "    # Find all releases\n",
    "    releases = soup.body.findAll(text=re.compile(\"release-\"))\n",
    "    # Get release numbers\n",
    "    rels = []\n",
    "    for rel in releases:\n",
    "        rels.append(rel.split(\"/\")[0].split(\"-\")[-1])\n",
    "\n",
    "    # Find highest release number (= latest release)\n",
    "    ENS_rel = np.array(rels).astype(int).max()\n",
    "\n",
    "    return ENS_rel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "106"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_latest_ens_rel()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gget_species_options(release=None):\n",
    "    \"\"\"\n",
    "    Function to find all available species core databases for gget.\n",
    "\n",
    "    Args:\n",
    "    - release   Ensembl release for which the databases are fetched.\n",
    "                (Default: latest release.)\n",
    "\n",
    "    Returns list of available core databases.\n",
    "    \"\"\"\n",
    "    # Find latest Ensembl release\n",
    "    ENS_rel = find_latest_ens_rel()\n",
    "\n",
    "    # If release != None, use user-defined Ensembl release\n",
    "    if release != None:\n",
    "        # Do not allow user-defined release if it is higher than the latest release\n",
    "        if release > ENS_rel:\n",
    "            raise ValueError(\n",
    "                \"Defined Ensembl release number cannot be greater than latest release.\"\n",
    "            )\n",
    "        else:\n",
    "            ENS_rel = release\n",
    "\n",
    "    # Find all available databases\n",
    "    url = f\"http://ftp.ensembl.org/pub/release-{ENS_rel}/mysql/\"\n",
    "    html = requests.get(url)\n",
    "\n",
    "    # Raise error if status code not \"OK\" Response\n",
    "    if html.status_code != 200:\n",
    "        raise RuntimeError(\n",
    "            f\"The Ensembl server returned error status code {html.status_code}. Please try again.\"\n",
    "        )\n",
    "\n",
    "    soup = BeautifulSoup(html.text, \"html.parser\")\n",
    "\n",
    "    # Return list of all available databases\n",
    "    databases = []\n",
    "    for subsoup in soup.body.findAll(\"a\"):\n",
    "        if \"core\" in subsoup[\"href\"]:\n",
    "            databases.append(subsoup[\"href\"].split(\"/\")[0])\n",
    "\n",
    "    return databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gget_species_options(release=2000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ref_species_options(which, release=None):\n",
    "    \"\"\"\n",
    "    Function to find all available species for gget ref.\n",
    "\n",
    "    Args:\n",
    "    - release   Ensembl release for which available species should be fetched.\n",
    "    - which     Which type of FTP. Possible entries: 'dna', 'cdna', 'gtf'.\n",
    "\n",
    "    Returns list of available species.\n",
    "    \"\"\"\n",
    "    # Find latest Ensembl release\n",
    "    ENS_rel = find_latest_ens_rel()\n",
    "\n",
    "    # If release != None, use user-defined Ensembl release\n",
    "    if release != None:\n",
    "        # Do not allow user-defined release if it is higher than the latest release\n",
    "        if release > ENS_rel:\n",
    "            raise ValueError(\n",
    "                \"Defined Ensembl release number cannot be greater than latest release.\"\n",
    "            )\n",
    "        else:\n",
    "            ENS_rel = release\n",
    "\n",
    "    # Find all available species for this release and FTP type\n",
    "    if which == \"gtf\":\n",
    "        url = f\"http://ftp.ensembl.org/pub/release-{ENS_rel}/gtf/\"\n",
    "    if which == \"dna\" or which == \"cdna\":\n",
    "        url = f\"http://ftp.ensembl.org/pub/release-{ENS_rel}/fasta/\"\n",
    "    html = requests.get(url)\n",
    "\n",
    "    # Raise error if status code not \"OK\" Response\n",
    "    if html.status_code != 200:\n",
    "        raise RuntimeError(\n",
    "            f\"The Ensembl server returned error status code {html.status_code}. Please try again.\"\n",
    "        )\n",
    "\n",
    "    # Parse the html and generate a clean list of the available genomes\n",
    "    soup = BeautifulSoup(html.text, \"html.parser\")\n",
    "\n",
    "    sps = []\n",
    "    for subsoup in soup.body.findAll(\"a\"):\n",
    "        sps.append(subsoup[\"href\"].split(\"/\")[0])\n",
    "\n",
    "    species_list = sps[1:]\n",
    "\n",
    "    # Return list of all available species\n",
    "    return species_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['acanthochromis_polyacanthus',\n",
       " 'accipiter_nisus',\n",
       " 'ailuropoda_melanoleuca',\n",
       " 'amazona_collaria',\n",
       " 'amphilophus_citrinellus',\n",
       " 'amphiprion_ocellaris',\n",
       " 'amphiprion_percula',\n",
       " 'anabas_testudineus',\n",
       " 'anas_platyrhynchos',\n",
       " 'anas_platyrhynchos_platyrhynchos',\n",
       " 'anas_zonorhyncha',\n",
       " 'anolis_carolinensis',\n",
       " 'anser_brachyrhynchus',\n",
       " 'anser_cygnoides',\n",
       " 'aotus_nancymaae',\n",
       " 'apteryx_haastii',\n",
       " 'apteryx_owenii',\n",
       " 'apteryx_rowi',\n",
       " 'aquila_chrysaetos_chrysaetos',\n",
       " 'astatotilapia_calliptera',\n",
       " 'astyanax_mexicanus',\n",
       " 'astyanax_mexicanus_pachon',\n",
       " 'athene_cunicularia',\n",
       " 'balaenoptera_musculus',\n",
       " 'betta_splendens',\n",
       " 'bison_bison_bison',\n",
       " 'bos_grunniens',\n",
       " 'bos_indicus_hybrid',\n",
       " 'bos_mutus',\n",
       " 'bos_taurus',\n",
       " 'bos_taurus_hybrid',\n",
       " 'bubo_bubo',\n",
       " 'buteo_japonicus',\n",
       " 'caenorhabditis_elegans',\n",
       " 'cairina_moschata_domestica',\n",
       " 'calidris_pugnax',\n",
       " 'calidris_pygmaea',\n",
       " 'callithrix_jacchus',\n",
       " 'callorhinchus_milii',\n",
       " 'camarhynchus_parvulus',\n",
       " 'camelus_dromedarius',\n",
       " 'canis_lupus_dingo',\n",
       " 'canis_lupus_familiaris',\n",
       " 'canis_lupus_familiarisbasenji',\n",
       " 'canis_lupus_familiarisboxer',\n",
       " 'canis_lupus_familiarisgreatdane',\n",
       " 'capra_hircus',\n",
       " 'capra_hircus_blackbengal',\n",
       " 'carassius_auratus',\n",
       " 'carlito_syrichta',\n",
       " 'castor_canadensis',\n",
       " 'catagonus_wagneri',\n",
       " 'catharus_ustulatus',\n",
       " 'cavia_aperea',\n",
       " 'cavia_porcellus',\n",
       " 'cebus_capucinus',\n",
       " 'cercocebus_atys',\n",
       " 'cervus_hanglu_yarkandensis',\n",
       " 'chelonoidis_abingdonii',\n",
       " 'chelydra_serpentina',\n",
       " 'chinchilla_lanigera',\n",
       " 'chlorocebus_sabaeus',\n",
       " 'choloepus_hoffmanni',\n",
       " 'chrysemys_picta_bellii',\n",
       " 'chrysolophus_pictus',\n",
       " 'ciona_intestinalis',\n",
       " 'ciona_savignyi',\n",
       " 'clupea_harengus',\n",
       " 'colobus_angolensis_palliatus',\n",
       " 'corvus_moneduloides',\n",
       " 'cottoperca_gobio',\n",
       " 'coturnix_japonica',\n",
       " 'cricetulus_griseus_chok1gshd',\n",
       " 'cricetulus_griseus_crigri',\n",
       " 'cricetulus_griseus_picr',\n",
       " 'crocodylus_porosus',\n",
       " 'cyanistes_caeruleus',\n",
       " 'cyclopterus_lumpus',\n",
       " 'cynoglossus_semilaevis',\n",
       " 'cyprinodon_variegatus',\n",
       " 'cyprinus_carpio',\n",
       " 'cyprinus_carpio_germanmirror',\n",
       " 'cyprinus_carpio_hebaored',\n",
       " 'cyprinus_carpio_huanghe',\n",
       " 'danio_rerio',\n",
       " 'dasypus_novemcinctus',\n",
       " 'delphinapterus_leucas',\n",
       " 'denticeps_clupeoides',\n",
       " 'dicentrarchus_labrax',\n",
       " 'dipodomys_ordii',\n",
       " 'dromaius_novaehollandiae',\n",
       " 'drosophila_melanogaster',\n",
       " 'echeneis_naucrates',\n",
       " 'echinops_telfairi',\n",
       " 'electrophorus_electricus',\n",
       " 'eptatretus_burgeri',\n",
       " 'equus_asinus_asinus',\n",
       " 'equus_caballus',\n",
       " 'erinaceus_europaeus',\n",
       " 'erpetoichthys_calabaricus',\n",
       " 'erythrura_gouldiae',\n",
       " 'esox_lucius',\n",
       " 'falco_tinnunculus',\n",
       " 'felis_catus',\n",
       " 'ficedula_albicollis',\n",
       " 'fukomys_damarensis',\n",
       " 'fundulus_heteroclitus',\n",
       " 'gadus_morhua',\n",
       " 'gallus_gallus',\n",
       " 'gambusia_affinis',\n",
       " 'gasterosteus_aculeatus',\n",
       " 'geospiza_fortis',\n",
       " 'gopherus_agassizii',\n",
       " 'gopherus_evgoodei',\n",
       " 'gorilla_gorilla',\n",
       " 'gouania_willdenowi',\n",
       " 'haplochromis_burtoni',\n",
       " 'heterocephalus_glaber_female',\n",
       " 'heterocephalus_glaber_male',\n",
       " 'hippocampus_comes',\n",
       " 'homo_sapiens',\n",
       " 'hucho_hucho',\n",
       " 'ictalurus_punctatus',\n",
       " 'ictidomys_tridecemlineatus',\n",
       " 'jaculus_jaculus',\n",
       " 'junco_hyemalis',\n",
       " 'kryptolebias_marmoratus',\n",
       " 'labrus_bergylta',\n",
       " 'larimichthys_crocea',\n",
       " 'lates_calcarifer',\n",
       " 'laticauda_laticaudata',\n",
       " 'latimeria_chalumnae',\n",
       " 'lepidothrix_coronata',\n",
       " 'lepisosteus_oculatus',\n",
       " 'leptobrachium_leishanense',\n",
       " 'lonchura_striata_domestica',\n",
       " 'loxodonta_africana',\n",
       " 'lynx_canadensis',\n",
       " 'macaca_fascicularis',\n",
       " 'macaca_mulatta',\n",
       " 'macaca_nemestrina',\n",
       " 'malurus_cyaneus_samueli',\n",
       " 'manacus_vitellinus',\n",
       " 'mandrillus_leucophaeus',\n",
       " 'marmota_marmota_marmota',\n",
       " 'mastacembelus_armatus',\n",
       " 'maylandia_zebra',\n",
       " 'meleagris_gallopavo',\n",
       " 'melopsittacus_undulatus',\n",
       " 'meriones_unguiculatus',\n",
       " 'mesocricetus_auratus',\n",
       " 'microcebus_murinus',\n",
       " 'microtus_ochrogaster',\n",
       " 'mola_mola',\n",
       " 'monodelphis_domestica',\n",
       " 'monodon_monoceros',\n",
       " 'monopterus_albus',\n",
       " 'moschus_moschiferus',\n",
       " 'mus_caroli',\n",
       " 'mus_musculus',\n",
       " 'mus_musculus_129s1svimj',\n",
       " 'mus_musculus_aj',\n",
       " 'mus_musculus_akrj',\n",
       " 'mus_musculus_balbcj',\n",
       " 'mus_musculus_c3hhej',\n",
       " 'mus_musculus_c57bl6nj',\n",
       " 'mus_musculus_casteij',\n",
       " 'mus_musculus_cbaj',\n",
       " 'mus_musculus_dba2j',\n",
       " 'mus_musculus_fvbnj',\n",
       " 'mus_musculus_lpj',\n",
       " 'mus_musculus_nodshiltj',\n",
       " 'mus_musculus_nzohlltj',\n",
       " 'mus_musculus_pwkphj',\n",
       " 'mus_musculus_wsbeij',\n",
       " 'mus_pahari',\n",
       " 'mus_spicilegus',\n",
       " 'mus_spretus',\n",
       " 'mustela_putorius_furo',\n",
       " 'myotis_lucifugus',\n",
       " 'myripristis_murdjan',\n",
       " 'naja_naja',\n",
       " 'nannospalax_galili',\n",
       " 'neogobius_melanostomus',\n",
       " 'neolamprologus_brichardi',\n",
       " 'neovison_vison',\n",
       " 'nomascus_leucogenys',\n",
       " 'notamacropus_eugenii',\n",
       " 'notechis_scutatus',\n",
       " 'nothobranchius_furzeri',\n",
       " 'nothoprocta_perdicaria',\n",
       " 'numida_meleagris',\n",
       " 'ochotona_princeps',\n",
       " 'octodon_degus',\n",
       " 'oncorhynchus_kisutch',\n",
       " 'oncorhynchus_mykiss',\n",
       " 'oncorhynchus_tshawytscha',\n",
       " 'oreochromis_aureus',\n",
       " 'oreochromis_niloticus',\n",
       " 'ornithorhynchus_anatinus',\n",
       " 'oryctolagus_cuniculus',\n",
       " 'oryzias_javanicus',\n",
       " 'oryzias_latipes',\n",
       " 'oryzias_latipes_hni',\n",
       " 'oryzias_latipes_hsok',\n",
       " 'oryzias_melastigma',\n",
       " 'oryzias_sinensis',\n",
       " 'otolemur_garnettii',\n",
       " 'otus_sunia',\n",
       " 'ovis_aries',\n",
       " 'ovis_aries_rambouillet',\n",
       " 'pan_paniscus',\n",
       " 'pan_troglodytes',\n",
       " 'panthera_leo',\n",
       " 'panthera_pardus',\n",
       " 'panthera_tigris_altaica',\n",
       " 'papio_anubis',\n",
       " 'parambassis_ranga',\n",
       " 'paramormyrops_kingsleyae',\n",
       " 'parus_major',\n",
       " 'pavo_cristatus',\n",
       " 'pelodiscus_sinensis',\n",
       " 'pelusios_castaneus',\n",
       " 'periophthalmus_magnuspinnatus',\n",
       " 'peromyscus_maniculatus_bairdii',\n",
       " 'petromyzon_marinus',\n",
       " 'phascolarctos_cinereus',\n",
       " 'phasianus_colchicus',\n",
       " 'phocoena_sinus',\n",
       " 'physeter_catodon',\n",
       " 'piliocolobus_tephrosceles',\n",
       " 'podarcis_muralis',\n",
       " 'poecilia_formosa',\n",
       " 'poecilia_latipinna',\n",
       " 'poecilia_mexicana',\n",
       " 'poecilia_reticulata',\n",
       " 'pogona_vitticeps',\n",
       " 'pongo_abelii',\n",
       " 'procavia_capensis',\n",
       " 'prolemur_simus',\n",
       " 'propithecus_coquereli',\n",
       " 'pseudonaja_textilis',\n",
       " 'pteropus_vampyrus',\n",
       " 'pundamilia_nyererei',\n",
       " 'pygocentrus_nattereri',\n",
       " 'rattus_norvegicus',\n",
       " 'rhinolophus_ferrumequinum',\n",
       " 'rhinopithecus_bieti',\n",
       " 'rhinopithecus_roxellana',\n",
       " 'saccharomyces_cerevisiae',\n",
       " 'saimiri_boliviensis_boliviensis',\n",
       " 'salarias_fasciatus',\n",
       " 'salmo_salar',\n",
       " 'salmo_trutta',\n",
       " 'salvator_merianae',\n",
       " 'sander_lucioperca',\n",
       " 'sarcophilus_harrisii',\n",
       " 'sciurus_vulgaris',\n",
       " 'scleropages_formosus',\n",
       " 'scophthalmus_maximus',\n",
       " 'serinus_canaria',\n",
       " 'seriola_dumerili',\n",
       " 'seriola_lalandi_dorsalis',\n",
       " 'sinocyclocheilus_anshuiensis',\n",
       " 'sinocyclocheilus_grahami',\n",
       " 'sinocyclocheilus_rhinocerous',\n",
       " 'sorex_araneus',\n",
       " 'sparus_aurata',\n",
       " 'spermophilus_dauricus',\n",
       " 'sphaeramia_orbicularis',\n",
       " 'sphenodon_punctatus',\n",
       " 'stachyris_ruficeps',\n",
       " 'stegastes_partitus',\n",
       " 'strigops_habroptila',\n",
       " 'strix_occidentalis_caurina',\n",
       " 'struthio_camelus_australis',\n",
       " 'suricata_suricatta',\n",
       " 'sus_scrofa',\n",
       " 'sus_scrofa_bamei',\n",
       " 'sus_scrofa_berkshire',\n",
       " 'sus_scrofa_hampshire',\n",
       " 'sus_scrofa_jinhua',\n",
       " 'sus_scrofa_landrace',\n",
       " 'sus_scrofa_largewhite',\n",
       " 'sus_scrofa_meishan',\n",
       " 'sus_scrofa_pietrain',\n",
       " 'sus_scrofa_rongchang',\n",
       " 'sus_scrofa_tibetan',\n",
       " 'sus_scrofa_usmarc',\n",
       " 'sus_scrofa_wuzhishan',\n",
       " 'taeniopygia_guttata',\n",
       " 'takifugu_rubripes',\n",
       " 'terrapene_carolina_triunguis',\n",
       " 'tetraodon_nigroviridis',\n",
       " 'theropithecus_gelada',\n",
       " 'tupaia_belangeri',\n",
       " 'tursiops_truncatus',\n",
       " 'urocitellus_parryii',\n",
       " 'ursus_americanus',\n",
       " 'ursus_maritimus',\n",
       " 'ursus_thibetanus_thibetanus',\n",
       " 'varanus_komodoensis',\n",
       " 'vicugna_pacos',\n",
       " 'vombatus_ursinus',\n",
       " 'vulpes_vulpes',\n",
       " 'xenopus_tropicalis',\n",
       " 'xiphophorus_couchianus',\n",
       " 'xiphophorus_maculatus',\n",
       " 'zalophus_californianus',\n",
       " 'zonotrichia_albicollis',\n",
       " 'zosterops_lateralis_melanops']"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_species_options(\"gtf\", 105)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_blast_ref_page(handle):\n",
    "    \"\"\"\n",
    "    Extract RID and RTOE from the NCBI 'please wait' page (handle).\n",
    "    RTOE = 'Estimated time fo completion.'\n",
    "    RID = 'Request ID'.\n",
    "\n",
    "    Returns RID, RTOE\n",
    "\n",
    "    Code partly adapted from the Biopython BLAST NCBIWWW project written\n",
    "    by Jeffrey Chang (Copyright 1999), Brad Chapman, and Chris Wroe distributed under the\n",
    "    Biopython License Agreement and BSD 3-Clause License\n",
    "    https://github.com/biopython/biopython/blob/171697883aca6894f8367f8f20f1463ce7784d0c/LICENSE.rst\n",
    "    \"\"\"\n",
    "\n",
    "    # Decode handle\n",
    "    string = handle.read().decode()\n",
    "\n",
    "    # Find RID\n",
    "    idx = string.find(\"RID =\")\n",
    "    if idx == -1:\n",
    "        rid = None\n",
    "    else:\n",
    "        jdx = string.find(\"\\n\", idx)\n",
    "        rid = string[idx + len(\"RID =\") : jdx].strip()\n",
    "\n",
    "    # Find RTOE\n",
    "    rtoe_idx = string.find(\"RTOE =\")\n",
    "    if rtoe_idx == -1:\n",
    "        rtoe = None\n",
    "    else:\n",
    "        rtoe_jdx = string.find(\"\\n\", rtoe_idx)\n",
    "        rtoe = string[rtoe_idx + len(\"RTOE =\") : rtoe_jdx].strip()\n",
    "\n",
    "    # If neither RID, nor RTOE were found, try to extract error message from HTML page\n",
    "    if not rid and not rtoe:\n",
    "        # Search for 'error msInf' div class\n",
    "        i = string.find('<div class=\"error msInf\">')\n",
    "        if i != -1:\n",
    "            msg = string[i + len('<div class=\"error msInf\">') :].strip()\n",
    "            msg = msg.split(\"</div>\", 1)[0].split(\"\\n\", 1)[0].strip()\n",
    "            if msg:\n",
    "                raise ValueError(f\"Error message from NCBI: {msg}\")\n",
    "        # Search for 'error' class\n",
    "        i = string.find('<p class=\"error\">')\n",
    "        if i != -1:\n",
    "            msg = string[i + len('<p class=\"error\">') :].strip()\n",
    "            msg = msg.split(\"</p>\", 1)[0].split(\"\\n\", 1)[0].strip()\n",
    "            if msg:\n",
    "                raise ValueError(f\"Error message from NCBI: {msg}\")\n",
    "        # Generic search for error messages\n",
    "        i = string.find(\"Message ID#\")\n",
    "        if i != -1:\n",
    "            # Break the message at the first HTML tag\n",
    "            msg = string[i:].split(\"<\", 1)[0].split(\"\\n\", 1)[0].strip()\n",
    "            raise ValueError(f\"Error message from NCBI: {msg}\")\n",
    "        # Raise general error, if the error layout was not recognized\n",
    "        raise ValueError(\n",
    "            \"No request ID and no estimated time to completion were found in the NCBI 'please wait' page.\"\n",
    "        )\n",
    "    # Raise error if RTOE was found but RID was not\n",
    "    elif not rid:\n",
    "        raise ValueError(\n",
    "            f\"No request ID (RID) was found in the NCBI 'please wait' page. (Although estimated time to completion = {rtoe}.)\"\n",
    "        )\n",
    "    # Raise error if RTOE was found but RID was not\n",
    "    elif not rtoe:\n",
    "        raise ValueError(\n",
    "            f\"No estimated time to completion was found in the NCBI 'please wait' page. (Although request ID = {rid}.)\"\n",
    "        )\n",
    "\n",
    "    try:\n",
    "        return rid, int(rtoe)\n",
    "    except ValueError:\n",
    "        raise ValueError(\n",
    "            f\"A non-integer estimated time to completion was found in the NCBI 'please wait' page: '{rtoe}'.\"\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "dd19f4a025a239cd738138a9cfd973ebc04741b3d7e2339b3ee95d5c6fc3738f"
  },
  "kernelspec": {
   "display_name": "Python 3.8.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
